### PR TITLE
231 add extract shared lines

### PR DESCRIPTION
**Pull request by tomiturunen1**
Adding function Extract Shared Lines between 2 polygons

### CODE REVIEW

**Comment by lehtonenp** [comment]
Conda / Conda Install & Test (3.9, windows-latest) fails. This is due to an error in the tests. Wait until https://github.com/GispoCoding/eis_toolkit/pull/240 is completed and merge master to your branch. I will check the code from your part.

**Comment by nmaarnio** [comment]
PR #240 has been now merged to master

File: eis_toolkit/vector_processing/extract_shared_lines.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def _extract_shared_lines(polygons: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    shared_lines_list = []
+
+    # Used to make sure that same polygon couples are not examined together multiple times
+    reversed_couples = set()
+
+    for i, poly1 in polygons.iterrows():
+        for j, poly2 in polygons.iterrows():
```
Comment (lehtonenp) [inline eis_toolkit/vector_processing/extract_shared_lines.py:14]: It is an option to use itertools.combinations to replace the for loops with just one. Then we would also not need to compare i to j.

File: eis_toolkit/vector_processing/extract_shared_lines.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import geopandas as gpd
+from beartype import beartype
+
+from eis_toolkit import exceptions
+
+
+def _extract_shared_lines(polygons: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    shared_lines_list = []
```
Comment (lehtonenp) [inline eis_toolkit/vector_processing/extract_shared_lines.py:8]: This _could_ be defined as `shared_lines_list: Sequence = []`

Comment (nmaarnio) [reply eis_toolkit/vector_processing/extract_shared_lines.py:8]: No need to define types for variables, except in maybe some rare cases.

**Comment by lehtonenp** [comment]
In the file `testing_extract_shared_lines.ipynb` you import `matplotlib.pyplot as plt` several times. This can be done one the first cell of the notebook once. :)

**Comment by nialov** [comment]
I think others have this review in check but a small suggestion regarding optimisation/simplification:

~~~python

# iterrows is usually not needed, I would just enumerate over the geometry
# column values
for i, poly in enumerate(polygons.geometry.values):
    
    # Select all other polygons except current
    index_set = set(polygons.index)
    index_set.remove(i)
    other_polygons = polygons.iloc[list(index_set)].geometry

    # Get intersection between the polygon and all other polygons
    shared_lines = other_polygons.intersection(poly)

    # Check that the intersection result is a line and does not yet exist in shared_lines_list
    for shared_line in shared_lines:
        if not shared_line.is_empty and (
            shared_line.geom_type == "MultiLineString" or shared_line.geom_type == "LineString"
        ) and shared_line not in shared_lines_list:
            shared_lines_list.append(shared_line)
~~~

Did a short test and I believe my version has the problem that the reverse lines will also be included in `shared_lines_list`.

Anyway, if you have the interest you could try this approach although I do not think it is necessary if your implementation works well!

**Review by Mtk112 — APPROVED** [review]
Everything looks good to me. Ready to be merged!

File: eis_toolkit/vector_processing/extract_shared_lines.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    return shared_lines_gdf
+
+
+@beartype
+def extract_shared_lines(polygons: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    """Extract shared lines/borders/edges between polygons.
+
+    Args:
+        polygons: The geodataframe that contains the polygon geometries to be examined
+                    for shared lines.
```
Comment (nmaarnio) [inline eis_toolkit/vector_processing/extract_shared_lines.py:36]: Just a minor suggestion so docstring looks the same as in other functions.
```suggestion
    Args:
        polygons: The geodataframe that contains the polygon geometries to be examined
            for shared lines.
```

File: eis_toolkit/vector_processing/extract_shared_lines.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    Returns:
+        Geodataframe containing the shared lines that were found between the polygons
```
Comment (nmaarnio) [inline eis_toolkit/vector_processing/extract_shared_lines.py:39]: Add a `Raises` section after the `Returns.`

**Review by nmaarnio — COMMENTED** [review]
Just a few modifications to the docstring and it's ready to go! I don't know if you've tried how fast this tool runs on a larger dataset, but we can leave that for testers too and optimize later if needed.

**Comment by nmaarnio** [comment]
Looks good now! Merging

**Pull request merged**